core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -215,24 +226,48 @@ private void acquireConnectionAndExecute0(ClientRequestContext ctx, Endpoint end
         }
     }
 
-    private ProxyConfig getProxyConfig(SessionProtocol protocol, Endpoint endpoint) {
+    private void resolveProxyConfig(SessionProtocol protocol, Endpoint endpoint, ClientRequestContext ctx,
+                                    BiConsumer<@Nullable ProxyConfig, @Nullable Throwable> onComplete) {
         final ProxyConfig proxyConfig = factory.proxyConfigSelector().select(protocol, endpoint);
         requireNonNull(proxyConfig, "proxyConfig");
 
+        final ServiceRequestContext serviceCtx = ServiceRequestContext.currentOrNull();
+        final ProxiedAddresses capturedProxiedAddresses = serviceCtx == null ? null
+                                                          : serviceCtx.proxiedAddresses();
+
+        // DirectProxyConfig does not have proxyAddress as its field.
+        if (proxyConfig.proxyAddress() != null) {
```

Comment:
Could you also skip IP address resolving if `proxyConfig.proxyAddress().isUnresolved() == false`?

Reply:
If I remember correctly, Netty AddressResolver returns the cached IP addresses if `InetSocketAddress` is resolved.

What do you think of adding methods that take a hostname as a string. We may create unsolved `InetSocketAddress` internally.

Reply:
@ikhoon nim, thanks for your comment! 🙇‍♂️ 
As you mentioned, It seems that `AddressResolve` works that way.

So, I think to add an private method taking a hostname as a string, is good way to refresh DNS. 
I made a new commit for it!

When you have time, please take another look. 🙇‍♂️

Reply:
> So, I think to add an private method taking a hostname as a string, is good way to refresh DNS.

I meant to add public factory methods to take a hostname. For example:
```java
public static Socks4ProxyConfig socks4(String proxyHostname, int proxyPort) {
    requireNonNull(proxyHostname, "proxyHostname");
    return socks4(InetSocketAddress.createUnresolved(proxyHostname, proxyPort));
}

public static Socks4ProxyConfig socks4(InetSocketAddress proxyAddress) {
    requireNonNull(proxyAddress, "proxyAddress");
    return new Socks4ProxyConfig(proxyAddress, null);
}
```

Reply:
@ikhoon nim, thanks for your comments. 🙇‍♂️ 
I see. 
I have a question.
Where do you think this factory method will be used?

If you expect this method to be used within `HttpClientDelegate#resolveProxyConfig(...)`, the private method `createUnresolvedAddressForRefreshing` that I wrote might be a good approach.

If we use for refreshing DNS in `HttpClientDelegate#resolveProxyConfig(...)`, the code complexity seems to increase significantly.
1. each factory methods have different method name (socks4, socks5, connect).
2. each factory methods have different method signature.
  - `socks4()`
    - `proxyAddress`
    - `proxyAddress, username`
  - `socks5()`
    - `proxyAddress`
    - `proxyAddress, username, password`
  - `connect()`
    - `proxyAddress, useTls`
    - `proxyAddres, username, password, useTls`
3. `HttpClientDelegate` know only `ProxyConfig` type. 

For the these reasons, we should add switch + If syntax to find type and select proper factory method in `HttpClientDelegate#resolveProxyConfig()`. 

We only refresh proxy address, IMHO, We only need to refresh the proxyAddress, so creating an unresolved `InetSocketAddress` and passing it to the ProxyConfig would make the code more readable.

What do you think about this? 
Or do you have any other concerns?

Reply:
> so creating an unresolved InetSocketAddress and passing it to the ProxyConfig 

Nevermind. I missed this part. I still prefer to add new factory methods for convenience and a clearer resolution policy, but they are less relevant to this PR. Let's address that as a separate issue if necessary.